File: Machine Learning A-Z/Part 2 - Regression/Section 6 - Polynomial Regression/Polynomial_Regression/polynomial_2.R
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[,2:3]

lin_regression = lm(Salary ~ Level, data = dataset)
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4


poly_lin_regression = lm(Salary ~ ., data = dataset)
summary(poly_lin_regression)

library(ggplot2)
#plotting linear regression graph
ggplot()+
  geom_point(aes(x =dataset$Level , y =dataset$Salary), color = 'red')+
  geom_line(aes(x = dataset$Level, y = predict(lin_regression, newdata = dataset)), color = 'blue')+
  ggtitle('Truth or Bluff (Linear Regression)') +
  xlab('Level') +
  ylab('Salary')

#plotting poly linear regression graph
ggplot()+
  geom_point(aes(x =dataset$Level , y =dataset$Salary), color = 'red')+
  geom_line(aes(x = dataset$Level, y = predict(poly_lin_regression, newdata = dataset)), color = 'blue')+
  ggtitle('Truth or Bluff (Polynomial Regression)') +
  xlab('Level') +
  ylab('Salary')


#smoother graph plot
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)
newDataSet = data.frame(Level = x_grid, 
                        Level2 = x_grid^2,
                        Level3 = x_grid^3,
                        Level4 = x_grid^4)
ggplot()+
  geom_point(aes(x =dataset$Level , y =dataset$Salary), color = 'red')+
  geom_line(aes(x = x_grid, y = predict(poly_lin_regression, newdata = newDataSet)), color = 'blue')+
  ggtitle('Truth or Bluff (Polynomial Regression)') +
  xlab('Level') +
  ylab('Salary')

#testing data for prediction
predict(lin_regression, data.frame(Level=6.5))
predict(poly_lin_regression, data.frame(Level = 6.5, Level2 = 6.5^2,Level3 = 6.5^3,Level4 = 6.5^4))

File: Machine Learning A-Z/Part 2 - Regression/Section 6 - Polynomial Regression/Polynomial_Regression/polynomial_2.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sat Jul  8 12:09:28 2017

@author: tgupta2
"""
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:,1:2].values
y = dataset.iloc[:,2].values
                
#from sklearn.cross_validation import train_test_split
#X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

from sklearn.linear_model import LinearRegression
linear_reg= LinearRegression()
linear_reg.fit(X, y)
y_pred = linear_reg.predict(X)

from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4)
#creating a polynomial feature vectors
X_poly= poly_reg.fit_transform(X)

lin_reg = LinearRegression()
lin_reg.fit(X_poly, y)
y_pred_poly = lin_reg.predict(X_poly)

# Visualising the Linear Regression results
plt.scatter(X, y, color = 'red')
plt.plot(X, y_pred, color = 'blue')
plt.title('Truth or Bluff (Linear Regression)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

# Visualising the Polynomial Regression results
plt.scatter(X, y, color = 'red')
plt.plot(X, y_pred_poly, color = 'blue')
plt.title('Truth or Bluff (Poly Regression)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()


# Visualising the Polynomial Regression results (for higher resolution and smoother curve)
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape((len(X_grid), 1))
plt.scatter(X, y, color = 'red')
plt.plot(X_grid, lin_reg.predict(poly_reg.fit_transform(X_grid)), color = 'blue')
plt.title('Truth or Bluff (Polynomial Regression)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

# Predicting a new result with Linear Regression
linear_reg.predict(6.5)

# Predicting a new result with Polynomial Regression
lin_reg.predict(poly_reg.fit_transform(6.5))